src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +73,158 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
+    if (rv)
+    {
+        ApplyTerminalSettings(termios, signalForBreak, hasInteractiveChildren);
+    }
+    return rv;
+}
 
-    if (signalForBreak)
-        termios->c_lflag |= (uint32_t)ISIG;
-    else
-        termios->c_lflag &= (uint32_t)(~ISIG);
+static void ttou_handler(int signo)
+{
+    (void)signo;
+    g_receivedSigTtou = true;
+}
+
+static bool TcSetAttr(struct termios* t, bool blockIfBackground)
+{
+    if (g_consoleUninitialized)
+    {
+        // The application is exiting, we mustn't change terminal settings.
+        return true;
+    }
+
+    if (!blockIfBackground)
+    {
+        // When the process is running in background, changing terminal settings
+        // will stop it (default SIGTTOU action).
+        // We change SIGTTOU to to get EINTR instead of blocking.
+        struct sigaction action;
+        memset(&action, 0, sizeof(action));
+        action.sa_handler = ttou_handler;
+        int rvSigaction = sigaction(SIGTTOU, &action, NULL);
+        assert(rvSigaction == 0);
+
+        g_receivedSigTtou = false;
+    }
+
+    bool rv = tcsetattr(STDIN_FILENO, TCSANOW, t) >= 0;
+
+    if (!blockIfBackground)
+    {
+        if (!rv && errno == EINTR && g_receivedSigTtou)
+        {
+            // Operation failed because we are background
+            // pretend it went fine.
+            rv = true;
+        }
+
+        // Restore default SIGTTOU handler.
+        struct sigaction action;
+        memset(&action, 0, sizeof(action));
+        action.sa_handler = SIG_DFL;
+        int rvSigaction = sigaction(SIGTTOU, &action, NULL);
+        assert(rvSigaction == 0);
+    }
+
+    return rv;
 }
 
-// In order to support Console.ReadKey(intercept: true), we need to disable echo and canonical mode.
-// We have two main choices: do so for the entire app, or do so only while in the Console.ReadKey(true).
-// The former has a huge downside: the terminal is in a non-echo state, so anything else that runs
-// in the same terminal won't echo even if it expects to, e.g. using Process.Start to launch an interactive,
-// program, or P/Invoking to a native library that reads from stdin rather than using Console.  The second
-// also has a downside, in that any typing which occurs prior to invoking Console.ReadKey(true) will
-// be visible even though it wasn't supposed to be.  The downsides of the former approach are so large
-// and the cons of the latter minimal and constrained to the one API that we've chosen the second approach.
-// Thus, InitializeConsoleBeforeRead is called to set up the state of the console, then a read is done,
-// and then UninitializeConsoleAfterRead is called.
-void SystemNative_InitializeConsoleBeforeRead(uint8_t minChars, uint8_t decisecondsTimeout)
+void UninitializeConsole()
 {
-    struct termios newTermios;
-    if (tcgetattr(STDIN_FILENO, &newTermios) >= 0)
+    // This method is called on SIGQUIT/SIGINT from the signal dispatching thread
+    // and on atexit.
```

Comment:
I've tried some things and it seems the terminal resets the settings when the process terminated. I wonder if we (still) need UninitializeConsole?

Reply:
I remember testing a bunch of cases, too, and explicitly adding the atexit because there were some where it ended up mattering... but I honestly can't remember which.  I'd be hesitant to remove it at this point, though, unless we're absolutely sure it's unnecessary in all situations, all terminals, etc.